### PR TITLE
Unpin versions for CI tools

### PR DESCRIPTION
**Pull request by avylove**
In order to avoid stale tools and to catch issues early, CI tools should pull the latest version.

This removes version pinning for dev and CI tools and fixes typing errors raised in newer versions of mypy.

### CODE REVIEW

File: lisa/util/__init__.py
Comment (squirrelsc) [inline lisa/util/__init__.py:468]: This change may have regression. If the dest is int, bool, float or str, the changed logic will take src, instead of dest.

File: lisa/sut_orchestrator/azure/common.py
Comment (squirrelsc) [inline lisa/sut_orchestrator/azure/common.py:176]: The marketplace object is used to make sure the `self._marketplace` is a final state, instead of an intermedia state. The value may be run at the same time. There is a race condition to get incomplete object. So it needs the local variable `marketplace` to get the whole object set correctly, then assign to `self._marketplace`

File: lisa/sut_orchestrator/azure/common.py
Comment (squirrelsc) [inline lisa/sut_orchestrator/azure/common.py:148]: The `__post_init__` may not be called, when creating the object in code. I'm not sure the logic, but there was some reason to declare them like InitVar, instead of put in `__post_init__`.

Comment (squirrelsc) [reply lisa/sut_orchestrator/azure/common.py:148]: Please try if it works with an Azure run.

File: lisa/util/__init__.py
Comment (avylove) [reply lisa/util/__init__.py:468]: The behavior hasn't changed. In the old logic, if `dest` was an `int`, `bool`, `float` or `str`, then `result` would not be a `dict`, causing `result` to fall back to `src`. This was all just useless logic.

File: lisa/sut_orchestrator/azure/common.py
Comment (avylove) [reply lisa/sut_orchestrator/azure/common.py:148]: `__post_init__()` is called by the `__init__()` method generating the data classes. This is where instance variables like this should go. `InitVar` types are passed to `__post_init__()`, but there was no logic for them, so that implementation was incomplete. It looks like they were just done that way to make mypy happy, but the logic in mypy changed and it was causing errors.

Kicked of an azure test local and it ran fine

**Review by squirrelsc — APPROVED** [review]

**Comment by squirrelsc** [comment]
Let's merge it next Monday.

**Comment by LiliDeng** [comment]
@squirrelsc is it right time to merge it? any testing need to do done before merging it?

**Comment by squirrelsc** [comment]
> @squirrelsc is it right time to merge it? any testing need to do done before merging it?

It's ok to merge it today or next week.

**Review by squirrelsc — CHANGES_REQUESTED** [review]
We can upgrade versions of tool, but not unpin them. The CI pipeline is being broken by flake8 upgrading.

**Comment by avylove** [comment]
> We can upgrade versions of tool, but not unpin them. The CI pipeline is being broken by flake8 upgrading.

I don't see how the internal CI pipeline is related, the version used there hasn't changed. It's best to unpin so these things get caught early. Otherwise you have what we currently have with out-of-date validation tools.

**Comment by squirrelsc** [comment]
> > We can upgrade versions of tool, but not unpin them. The CI pipeline is being broken by flake8 upgrading.
> 
> I don't see how the internal CI pipeline is related, the version used there hasn't changed. It's best to unpin so these things get caught early. Otherwise you have what we currently have with out-of-date validation tools.

I haven't said it happened on internal CI pipelines. But if we use the same way to unpin upstream packages in internal pipelines, that will happen someday. And we need to comeback to discuss and investigate in holidays again. The flake8 issue is more like a regression, instead of a new feature. It blocks me to merge some PRs. Even flake8 brings a new validation, it applies to existing code as a debt, which should be fixed by maintainers, instead of current PRs' contributors. I have to wait it's fixed. Out-of-date validation tools are cost, which can be accepted, but build breaks are not.

**Comment by avylove** [comment]
> > > We can upgrade versions of tool, but not unpin them. The CI pipeline is being broken by flake8 upgrading.
> > 
> > 
> > I don't see how the internal CI pipeline is related, the version used there hasn't changed. It's best to unpin so these things get caught early. Otherwise you have what we currently have with out-of-date validation tools.
> 
> I haven't said it happened on internal CI pipelines. But if we use the same way to unpin upstream packages in internal pipelines, that will happen someday. And we need to comeback to discuss and investigate in holidays again. The flake8 issue is more like a regression, instead of a new feature. It blocks me to merge some PRs. Even flake8 brings a new validation, it applies to existing code as a debt, which should be fixed by maintainers, instead of current PRs' contributors. I have to wait it's fixed. Out-of-date validation tools are cost, which can be accepted, but build breaks are not.

It's all one in the same and unpinning has the lowest maintenance cost. What you're describing is signal, which is what we want. If someone submits a PR and gets and error unrelated to their PR, they should be able to identify it. We fix it quickly or pin temporarily. The alternative is we need to come up with a system to automatically generate PRs for the next version. It's too much work to do that way which is why the standard way is to have no upper bound pinning.

**Comment by avylove** [comment]
Closing this since it's now handled by BumpDeps.
Typing changes for newer versions of MyPY merged into https://github.com/microsoft/lisa/pull/2545

**Pull request closed without merging**
